lib/streamlit/caching/cache_utils.py

```diff
@@ -85,14 +138,51 @@ def cache_type(self) -> CacheType:
         raise NotImplementedError
 
     @property
-    def call_stack(self) -> "CachedFunctionCallStack":
+    def warning_call_stack(self) -> "CacheWarningCallStack":
+        raise NotImplementedError
+
+    @property
+    def message_call_stack(self) -> "CacheMessagesCallStack":
         raise NotImplementedError
 
     def get_function_cache(self, function_key: str) -> Cache:
         """Get or create the function cache for the given key."""
         raise NotImplementedError
 
 
+def replay_result_messages(result: CachedResult) -> None:
+    """Replay the st element function calls that happened when executing a cache-decorated function.
+
+    When a cache function is executed, we record the element and block messages produced,
+    and use those to reproduce the DeltaGenerator calls, so the elements will appear in
+    the web app even when execution of the function is skipped because the result was cached.
+
+    To make this work, for each st function call we record an identifier for the
+    DG it was effectively called on. We also record the identifier for each DG returned
+    by an st function call. Then, for each recorded message, we get the current DG
+    instance corresponding to the DG the message was originally called on, and enqueue
+    the message using that, recording any new DGs produced in case a later st function
+    call is on one of them.
```

Comment:
So it looks like this comment is saying that we record BOTH the identifier for the dg that was called on (does that mean st.foo?) and then the dg returned by an st function call (when does this happen?) and then we enqueue each of those messages again and in each of those messages, we keep track of any new dgs? I'm personally confused about this method.

Reply:
`st.foo` is implicitly a call on the main DG, but st functions can also be called on other DGs like the sidebar, a container, etc. The main reason a dg is returned from an st function is for layout blocks, but anything that doesn't have a real return value (like st.write, st.empty) will actually return a dg so you can do weird layout stuff with them